src/projections/spilhaus.cpp

```diff
@@ -0,0 +1,179 @@
+
+namespace { // anonymous namespace
+struct pj_spilhaus_data {
+    double cosalpha;
+    double sinalpha;
+    double beta;
+    double lambda_0;
+    double conformal_distortion;
+    double cosrot;
+    double sinrot;
+    PJ *adams_ws2;
+};
+
+double conformal_lat(double lat, double e) {
+    // Snyder, A working manual, formula 3-1
+    const double esinlat = e * sin(lat);
+    const double chi =
+        2 * atan(tan(M_FORTPI + lat / 2) *
+                 std::pow(((1 - esinlat) / (1 + esinlat)), (e / 2))) -
+        M_HALFPI;
+    return chi;
+}
+} // anonymous namespace
+
+static PJ_XY spilhaus_s_forward(PJ_LP lp, PJ *P) {
+    PJ_XY xy = {0.0, 0.0};
+    struct pj_spilhaus_data *Q =
+        static_cast<struct pj_spilhaus_data *>(P->opaque);
+
+    const double phi_c = conformal_lat(lp.phi, P->e);
+    const double cosphi_c = cos(phi_c);
+    const double sinphi_c = sin(phi_c);
+
+    const double coslam = cos(lp.lam - Q->lambda_0);
+    const double sinlam = sin(lp.lam - Q->lambda_0);
+
+    PJ_LP lpadams;
+    // Snyder, A working manual, formula 5-7
+    lpadams.phi =
+        aasin(P->ctx, Q->sinalpha * sinphi_c - Q->cosalpha * cosphi_c * coslam);
+
+    // Snyder, A working manual, formula 5-8b
+    lpadams.lam =
+        (Q->beta + atan2(cosphi_c * sinlam, (Q->sinalpha * cosphi_c * coslam +
+                                             Q->cosalpha * sinphi_c)));
+
+    while (lpadams.lam > M_PI)
+        lpadams.lam -= M_PI * 2;
+    while (lpadams.lam < -M_PI)
+        lpadams.lam += M_PI * 2;
+
+    PJ_XY xyadams = Q->adams_ws2->fwd(lpadams, Q->adams_ws2);
+    const double factor = Q->conformal_distortion * P->k0;
+    xy.x = -(xyadams.x * Q->cosrot + xyadams.y * Q->sinrot) * factor;
+    xy.y = -(xyadams.x * -Q->sinrot + xyadams.y * Q->cosrot) * factor;
+
+    return (xy);
+}
+
+static PJ_LP spilhaus_s_inverse(PJ_XY xy, PJ *P) {
+    PJ_LP lp = {0.0, 0.0};
+    struct pj_spilhaus_data *Q =
+        static_cast<struct pj_spilhaus_data *>(P->opaque);
+
+    PJ_XY xyadams;
+    const double factor = 1.0 / (Q->conformal_distortion * P->k0);
+    xyadams.x = -(xy.x * Q->cosrot + xy.y * -Q->sinrot) * factor;
+    xyadams.y = -(xy.x * Q->sinrot + xy.y * Q->cosrot) * factor;
+    PJ_LP lpadams = Q->adams_ws2->inv(xyadams, Q->adams_ws2);
+
+    const double cosphi_s = cos(lpadams.phi);
+    const double sinphi_s = sin(lpadams.phi);
+    const double coslam_s = cos(lpadams.lam - Q->beta);
+    const double sinlam_s = sin(lpadams.lam - Q->beta);
+
+    // conformal latitude
+    lp.phi = aasin(P->ctx,
+                   Q->sinalpha * sinphi_s + Q->cosalpha * cosphi_s * coslam_s);
+
+    lp.lam = Q->lambda_0 +
+             aatan2(cosphi_s * sinlam_s,
+                    Q->sinalpha * cosphi_s * coslam_s - Q->cosalpha * sinphi_s);
+
+    // latitude (iterative formula from
+    // https://mathworld.wolfram.com/ConformalLatitude.html)
+    const double taninit = tan(M_PI / 4 + lp.phi / 2);
```

Comment:
could be worth moving in the same file as conformat_lat(), and find other users of it in the code base

Reply:
done